src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,208 @@
+    length(pswfcs) == length(pswfc_ang_moms) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "angular momenta do not agree.")
+    length(pswfcs) == length(pswfc_occs) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "occupations do not agree.")
+
+    return PspUpf{T}(Zion, lmax, rgrid, drgrid, vloc, projs, h,
+                     pswfcs, pswfc_ang_moms, pswfc_occs, rvlocpzdr, rprojsdr, identifier)
+end
+
+function parse_upf_file(path; identifier=path)
+    pseudo = load_upf(path)
+    # Pseudo-ion charge
+    Zion = Int(pseudo["header"]["z_valence"])
+    # Radial grid
+    rgrid = pseudo["radial_grid"]
+    drgrid = pseudo["radial_grid_derivative"]
+    # Maximum angular momentum channel
+    lmax = pseudo["header"]["l_max"]
+    # Number of Kleinman-Bylander β projectors
+    nproj = pseudo["header"]["number_of_proj"]
+    # Number of pseudo-atomic wavefunctions
+    nwfc = pseudo["header"]["number_of_wfc"]
+    # Local potential
+    vloc = pseudo["local_potential"]
+
+    # Kleinman-Bylander β projectors
+    projs = [Vector[] for _ = 0:lmax]
+    for i = 1:nproj
+        proj_data = pseudo["beta_projectors"][i]
+        l = proj_data["angular_momentum"]
+        proj = proj_data["radial_function"]
+        push!(projs[l+1], proj)
+    end
+    
+    # β-projector coupling coefficients
+    h = Matrix[]
+    n = 1
+    for l = 0:lmax
+        nproj_l = length(projs[l+1])
+        push!(h, pseudo["D_ion"][n:n+nproj_l-1, n:n+nproj_l-1] * 2)
+        n += nproj_l
+    end
+    
+    # Pseudo-atomic wavefunctions
+    pswfcs = Vector{Vector}(undef, nwfc)
+    pswfc_ang_moms = Vector(undef, nwfc)
+    pswfc_occs = Vector(undef, nwfc)
+    for i = 1:nwfc
+        pswfc_data = pseudo["atomic_wave_functions"][i]
+        pswfcs[i] = pswfc_data["radial_function"]
+        pswfc_ang_moms[i] = pswfc_data["angular_momentum"]
+        pswfc_occs[i] = pswfc_data["occupation"]
+    end
+
+    # Useful precomputed quantities
+    rvlocpzdr = (rgrid .* vloc .+ 2Zion) .* drgrid
+    rprojsdr = [Vector[] for _ = 0:lmax]
+    for l = 0:lmax
+        for proj in projs[l+1]
+            push!(rprojsdr[l+1], proj .* view(rgrid, 1:length(proj)) .* view(drgrid, 1:length(proj)))
+        end
+    end
+
+    PspUpf(Zion, lmax, rgrid, drgrid, vloc, projs, h,
+           pswfcs, pswfc_ang_moms, pswfc_occs, rvlocpzdr, rprojsdr; identifier)
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
+
+"""
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r::Number)
+
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space distance r
+via cubic spline interpolation on the real-space mesh.
+
+!!! Note
+UPFs store `rgrid[i] * β[l,n](rgrid[i])`, so we must divide by `r`.
+"""
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T) where {T <: Real}
+    ir_cut = length(psp.projs[l+1][i])
```

Comment:
Keep uniform notation ... I'd suggest to use `ir_cut` *or* `N` *or* sth else consistently everywhere.

Reply:
The local potential should always be the same size as the grid, while each projector is only read up until its value is approximately zero. This is mentioned in the documentation of the UPF format, where they mention numeric stability. I've found that the projectors are well-localized, so there's actually significant performance benefit to this as well. 

All this to say that the iteration limit means something different in the local potential and projectors. (Maybe `ir_max` and `ir_cut`?)